File: scala/src/main/scala/fun/pierre/lists/P06.scala
package fun.pierre.lists

import scala.annotation.tailrec


object P06 {

  @tailrec
  def isPalindrome[T](l: List[T]) : Boolean = {
    l match {
      case Nil => true
      case List(a) => true
      case list => list.head == list.last && isPalindrome(list.tail.init)
    }
  }
}

File: java/src/main/java/fun/pierre/lists/P06.java
package fun.pierre.lists;

import java.util.List;

public class P06 {

  public static<T> boolean isPalindrome(List<T> _list) {
    int size = _list.size();
    if (size <= 1) {
      return true;
    }
    return (_list.get(0) == _list.get(size - 1)) && isPalindrome(_list.subList(1, size - 1));
  }
}
